Show formatted dates in load/save window

AaronVanGeffen opened the pull request:
The load/save window has had the ability to sort by date since its inception, but it has never actually _shown_ the dates. This PR aims to change that.

The date format used is the one provided by the system locale settings. Hence, it looks different when comparing e.g. English to Japanese:

![2018-01-06 01 17 00](https://user-images.githubusercontent.com/604665/34633919-b37ccc98-f27f-11e7-9e8a-93df424ad2ac.png)  ![2018-01-06 01 22 54_](https://user-images.githubusercontent.com/604665/34634006-5eef2184-f280-11e7-8842-2a4e436bc640.png)

I'm not sure what to do about the date alignment. As the sprite font, like many Latin fonts, uses variable-width characters, it's a bit of a mess. A solution to this would be to possibly paint using two virtual columns, rendering the date right-aligned and the time left-aligned next to it?

Another question to ask is: does this make the window too small? Should we widen it a little?

AaronVanGeffen commented:
From Gitter, for posterity:

> 01:14 [marijnvdwerf] What about only showing a readable date in the list, and a full date on hover?
> 01:25 [AaronVanGeffen] By readable date you mean "two hours ago" and all?
> 01:26 [marijnvdwerf] Ideally: yes. But just a shorter date would suffice to start with. Time is less interesting when you want to open 'last weeks save'
> 01:32 [marijnvdwerf] I think Google drive uses that as well. If <24hrs ago: time. Otherwise date
> 01:32 [marijnvdwerf] Not overcomplicated, yet gives a decent idea without being too vague

Gymnasiast commented:
This would be a good time to make this window resizeable: as you can see it's starting to get cramped in there, and you have quite short save names.

AaronVanGeffen commented:
Yes, that's why I asked whether to wide it to begin with. Making it resizeable sounds like a good idea, too.

Gymnasiast commented:
Oh yes, you can make it wider by default as well.

rwjuk commented:
I'd probably go for two virtual columns to keep the dates & times aligned, but not if it will create an unreasonable amount of work and/or maintenance burden.

AaronVanGeffen commented:
Alright, I have implemented virtual columns for alignment. I'm not convinced that it looks nice, though.

The column should probably be resized based on the maximum date width for the current locale (e.g. '0000-00-00 00:00:00'), rather than having it fixed width and date/time offset in the middle of the column.

![screen shot 2018-01-07 at 14 57 45](https://user-images.githubusercontent.com/604665/34650099-616f2210-f3bb-11e7-8504-d1cf5f6b4923.png) ![screen shot 2018-01-07 at 14 55 58](https://user-images.githubusercontent.com/604665/34650098-615649c0-f3bb-11e7-8fc6-194c7dbbc19b.png)

Oh, and the window is resizeable now. The date column has a fixed width.

![screen shot 2018-01-07 at 14 57 52](https://user-images.githubusercontent.com/604665/34650100-618680cc-f3bb-11e7-8ee4-eece2d52bff6.png)

Ideas or suggestions for improvement are welcome.

AaronVanGeffen commented:
Alright, I have implemented virtual columns for alignment. I've also changed the headers to be mor header-like. 

The date column width is set to an approximation of the maximum date width for the current locale by computing the width for '2000-10-20 00:00:00'.

For English, this looks like:
![screen shot 2018-01-07 at 17 49 25](https://user-images.githubusercontent.com/604665/34651701-9d19e5f8-f3d3-11e7-8705-48c1f17cd065.png)

And for Japanese:
![screen shot 2018-01-07 at 17 48 29](https://user-images.githubusercontent.com/604665/34651700-9d00a462-f3d3-11e7-8dfd-a7f6c5c268be.png)
(For some reason, my locale doesn't specify dates to be formatted like 2000年10月9日...)

Oh, and the window is resizeable now. That's pretty cool.
![screen shot 2018-01-07 at 17 49 44](https://user-images.githubusercontent.com/604665/34651702-9d30fcca-f3d3-11e7-8849-421bea99fc07.png)

Suggestions for improvement are welcome.

marijnvdwerf commented:
Do all locales use the `date` `time` order?

AaronVanGeffen commented:
The order of both is not defined by the locales. We've opted to use `%x` on the left (date), and `%X` on the right (time).

marijnvdwerf commented:
It doesn't really display the file name though, as no extensions appear to be shown.

marijnvdwerf commented:
What happens when your system is set to Japanese, and the game is set to English?

rwjuk commented:
![image](https://user-images.githubusercontent.com/16639257/34655560-88e2ff22-f403-11e7-96f5-808d176283f1.png)

Screenshot above was taken with time set to 12-hour format @AaronVanGeffen

AaronVanGeffen commented:
@rwjuk It seems to be missing the a.m./p.m. indicator. Did you include that in your locale settings?

rwjuk commented:
@AaronVanGeffen Fixed!

rwjuk commented on src/openrct2/platform/Platform.Win32.cpp:
```diff
@@ -156,25 +156,27 @@ namespace Platform
-        SYSTEMTIME * st = TimeToSystemTime(timestamp);
+        SYSTEMTIME st = TimeToSystemTime(timestamp);
 
         wchar_t date[20];
```
Do we know that 20 bytes is enough for all locales?

AaronVanGeffen replied:
Well, a `wchar_t` is typically UTF-16 or even UTF-32, so it should be enough for basic representations. It might _not_ be enough if the short date format is set to print the full name of days, though — "Wednesday" is be 9 characters already. But I don't think that is very likely to happen…?

IntelOrca commented on src/openrct2/platform/Platform.Win32.cpp:
```diff
@@ -144,6 +145,39 @@ namespace Platform
         return path;
     }
+
+    SYSTEMTIME TimeToSystemTime(std::time_t timestamp)
```
static

AaronVanGeffen commented on src/openrct2-ui/windows/LoadSave.cpp:
```diff
@@ -711,6 +817,11 @@ static void window_loadsave_populate_list(rct_window *w, sint32 includeNewItem,
                 listItem->date_modified = platform_file_get_modified_time(listItem->path);
+
+                // Cache a human-readable version of the modified date.
+                listItem->date_formatted = Platform::FormatShortDate(listItem->date_modified);
```
These assignments seem to be problematic. I'd expect a copy constructor to kick in from the returned std::string, but that doesn't seem to be happening, resulting in instability.

AaronVanGeffen commented:
There seems to be something going wrong it the allocation process. Working hypothesis is that because `Memory::AllocateArray` allocates memory, but doesn't initialise it, the `std::string` constructor tries to call a destructor for an invalid pointer (garbage data) on `loadsave_list_item`.

I'm refactoring `_listItems` and `_listItemsCount` to use a `std::vector` instead of the C legacy data types.

IntelOrca commented on src/openrct2-ui/windows/LoadSave.cpp:
```diff
@@ -722,15 +723,15 @@ static bool list_item_sort(LoadSaveListItem &a, LoadSaveListItem &b)
     {
     case SORT_NAME_ASCENDING:
-        return strcicmp(a.name, b.name) < 0;
+        return a.name.compare(b.name) < 0;
```
You have lost the case insensitive comparison.

AaronVanGeffen replied:
Ah, you're right. Shall I revert it to use Util.h's `strcicmp` using C-strings, then?

AaronVanGeffen commented:
Okay, I've refactored the container into std::vector, and a bunch of C-strings into std::strings. I believe this is ready to be reviewed now; it's gotten out of hand enough as it is.

AaronVanGeffen commented:
This has been rebased to resolve conflicts resulting from #6925 being merged. As far as I'm concerned, this PR is done now. Before merging, I'd like to squash a few commits, though.

I'd appreciate people taking it out for a test run.

IntelOrca commented:
I just tried `b6eb365` and it still shows me mm/dd/yy 24h clock instead of dd/mm/yy 12h. I'll see if I can find out what the problem is.

tobiaskohlbau commented:
I've tried newest commit and got:

![image](https://user-images.githubusercontent.com/388354/34780043-4b976f2e-f622-11e7-939a-b55be800dcd2.png)

Looks kinda weird is it intended?

AaronVanGeffen commented:
@tobiaskohlbau Hmm, it is working properly for me. What did you do to get it in that state?

![screen shot 2018-01-10 at 16 29 09](https://user-images.githubusercontent.com/604665/34780443-74fbee5c-f623-11e7-810a-ed1ae02cfe8e.png)

tobiaskohlbau commented:
I've checked out `b6eb365` and build as win64, afterwards I've opened the load/save window and from standard size I minimized it to the smallest possible value. (Bad drawing starts as soon as first vertical minimizing appears.)

AaronVanGeffen commented:
@tobiaskohlbau Does 5416e2273297956355f8b626799d019879766577 fix the problem?

IntelOrca commented:
@AaronVanGeffen I have given up, I can't find anywhere, or any example in C++ to retrieve the user's custom time format for Windows. All the format functions want a locale name and the only name I can get back is `en-GB`, not my custom locale which has my custom time formats. In .NET it is very trivial, you just do `DateTime.ToShortDateString()` and you get back the custom format but looking through the .NET source code it was all very complicated.

This can be merged and if anyone knows how to resolve the issue they can just amend the platform function.

AaronVanGeffen commented:
@IntelOrca I'm sorry to hear it doesn't work properly... When @rwjuk tested it on Windows the other day, he specifically set the regional settings to use a 12-hour clock, producing the screenshot above.

With your permission, I'll merge it for now, then.

rwjuk commented:
Weird, I'll have another look and see what could be going on. Definitely displayed correctly for me...
